New version of upgrade.php

stefaweb opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mailwatch/1.2.0/441)
<!-- Reviewable:end -->

stefaweb commented:
To check.

- convert database to utf8
- fixes for tables
- convert database and tables to utf8mb4 if possible, unless convert tables to utf8

To do if necessary:
- detect if database and tables are already converted to utf8 or utf8mb4 (when we run more than one time upgrade.php)

stefaweb commented:
Result:

```
Testing connectivity to the database ................................. OK
Updating database schema:
 - Convert database to UTF-8.......................................... OK
 - Fix structure for username field in `users` table.................. OK
 - Fix structure for user field in `spamscores` table................. OK
 - Fix structure for username field in `audit_log` table.............. OK
 - Fix structure for password field in `users` table.................. OK
 - Fix structure for fullname field in `users` table.................. OK
 - Fix structure for username field in `user_filters` table........... OK
 - Fix structure for username field in `user_filters` table........... OK
 - Fix structure for username field in `user_filters` table........... OK
 - Fix structure for username field in `user_filters` table........... OK
 - Convert database to utf8mb4........................................ OK
 - Convert table `audit_log` to utf8mb4............................... ALREADY CONVERTED
 - Convert table `autorelease` to utf8mb4............................. ALREADY CONVERTED
 - Convert table `blacklist` to utf8mb4............................... ALREADY CONVERTED
 - Convert table `inq` to utf8mb4..................................... ALREADY CONVERTED
 - Convert table `maillog` to utf8mb4................................. ALREADY CONVERTED
 - Convert table `mtalog_ids` to utf8mb4.............................. ALREADY CONVERTED
 - Convert table `mcp_rules` to utf8mb4............................... ALREADY CONVERTED
 - Convert table `mtalog` to utf8mb4.................................. ALREADY CONVERTED
 - Convert table `outq` to utf8mb4.................................... ALREADY CONVERTED
 - Convert table `sa_rules` to utf8mb4................................ ALREADY CONVERTED
 - Convert table `saved_filters` to utf8mb4........................... ALREADY CONVERTED
 - Convert table `spamscores` to utf8mb4.............................. ALREADY CONVERTED
 - Convert table `user_filters` to utf8mb4............................ ALREADY CONVERTED
 - Convert table `users` to utf8mb4................................... ALREADY CONVERTED
 - Convert table `whitelist` to utf8mb4............................... ALREADY CONVERTED
 - Drop `geoip_country` table......................................... OK

Checking MailScanner.conf settings:
 - QuarantineWholeMessage ............................................ OK
 - QuarantineWholeMessagesAsQueueFiles ............................... OK
 - DetailedSpamReport ................................................ OK
 - IncludeScoresInSpamAssassinReport ................................. OK
 - SpamActions ....................................................... OK
 - HighScoringSpamActions ............................................ WARNING
 - AlwaysLookedUpLast ................................................ OK

*** ERROR/WARNING SUMMARY ***
MailScanner.conf: HighScoringSpamActions != store (=delete)
```

Skywalker-11 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-:-KbPAQQ9EHP7atllbsCD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Found a problem in the logic here. Tables not upgraded to utf8mb4.

if (check_utf8_table(DB_NAME, $table, $server_utf8_variant) === false) `{`

stefaweb commented:
![capture d ecran 2017-01-26 a 13 19 22](https://cloud.githubusercontent.com/assets/12907102/22331170/41b41450-e3ca-11e6-8e56-090fb48bdcc7.png)

![capture d ecran 2017-01-26 a 13 19 30](https://cloud.githubusercontent.com/assets/12907102/22331173/464e2910-e3ca-11e6-8cbd-9859aec70da0.png)

stefaweb commented:
Another problem.
When I force utf8 upgrade (changing `server_version >= 50503`), 2 tables are not converted.

![capture d ecran 2017-01-26 a 14 18 59](https://cloud.githubusercontent.com/assets/12907102/22332720/a1ba8c1e-e3d2-11e6-9338-9724e04be3b9.png)

stefaweb commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Something wrong somewhere.

With var_dump.

```
- Convert table `spamscores` to utf8.................................string(4) "utf8"
bool(false)
 OK
string(78) "ALTER TABLE `spamscores` CONVERT TO CHARACTER SET utf8 COLLATE utf8_unicode_ci"
 ```

When I execute the SQL command directly in phpmyadmin, the system answer OK but the table stay with utf8mb4.

Skywalker-11 commented:
hmm. here it works. what does this show

`SHOW CREATE TABLE spamscores;`

stefaweb commented:
mariadb-server 10.0.29+maria-1~jessie

Forced to utf8:
```
CREATE TABLE `spamscores` (
 `user` varchar(191) CHARACTER SET utf8 COLLATE utf8_unicode_ci NOT NULL DEFAULT '',
 `lowspamscore` decimal(10,0) NOT NULL DEFAULT '0',
 `highspamscore` decimal(10,0) NOT NULL DEFAULT '0',
 PRIMARY KEY (`user`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
```

I have the art of falling on problems...

Skywalker-11 commented:
the following still shows the last line with utf8mb4? I am using the the db version and here it works 

```
 ALTER TABLE `spamscores` CONVERT TO CHARACTER SET utf8 COLLATE utf8_unicode_ci;
SHOW CREATE TABLE spamscores;
```

stefaweb commented:
I tried from stratch again. This times it worked with upgrade.php. I don't understand.

Enclosed are the schema dump in zip to check.

First, I've the database, tables and fields all in utf8.
mailscanner-utf8.sql

Second, I run upgrade.php.
mailscanner-upgraded.sql

Third, I force back to utf8 (`server_version >= 650503`)
mailscanner-forcedback-utf8.sql
(this one not good but not really a problem)

Fourth, I run upgrade.php but say that mysql < 5.3 with `server_version >= 650503`
mailscanner-mysql5.3-utf8.sql

[databases.zip](https://github.com/mailwatch/1.2.0/files/732560/databases.zip)

stefaweb commented:
And the last.

With an "upgraded database" with upgrade.php, I run a second time upgrade.php.
This time, the script breaks the database...
[mailscanner-second-upgrade.sql.zip](https://github.com/mailwatch/1.2.0/files/732585/mailscanner-second-upgrade.sql.zip)

Seems, we need to add more test in the script.
If you have an idea.

stefaweb commented:
I found the problem. The ALTER for fields included CHARACTER SET and LOCATION. I removed them.
Also corrected some errors in schema. I compared with create.sql.

stefaweb commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

endelwar commented on upgrade.php:
```diff
@@ -130,36 +130,90 @@ function getTableIndexes($table)
+    $sql = "ALTER TABLE `user_filters` CHANGE `username` `username` VARCHAR( 191 ) NOT NULL DEFAULT ''";
+    executeQuery($sql);
+    
+    // Revert back some tables to the right values due to previous errors in upgrade.php
```
If username has to be 191 chars at max the "right" values is always 191, not more: every reference of username in tables has to be 191 chars long

stefaweb replied:
ok

endelwar commented on upgrade.php:
```diff
@@ -130,36 +130,90 @@ function getTableIndexes($table)
+
+    // Table autorelease
+    echo pad(' - Fix schema for msg_id field in `autorelease` table');
+    $sql = "ALTER TABLE `autorelease` CHANGE `msg_id` `msg_id` VARCHAR( 255 ) NOT NULL";
```
msg_id has always been 255 chars

stefaweb replied:
ok

endelwar commented on upgrade.php:
```diff
@@ -130,36 +130,90 @@ function getTableIndexes($table)
+    executeQuery($sql);
+
+    echo pad(' - Fix schema for uid field in `autorelease` table');
+    $sql = "ALTER TABLE `autorelease` CHANGE `uid` `uid` VARCHAR( 255 ) NOT NULL";
```
uid has always been 255 chars

stefaweb replied:
ok

endelwar commented on upgrade.php:
```diff
@@ -130,36 +130,90 @@ function getTableIndexes($table)
+    // Convert database to UTF-8mb4 if MySQL ≥ 5.5.3
+    if ($link->server_version >= 50503) {
+        $server_utf8_variant = 'utf8mb4';
+        echo pad(' - Convert database to ' . $server_utf8_variant . '');
```
lines 190-194 should be outside `if ($link->server_version >= 50503)` (always convert to utf8)

stefaweb replied:
Look at line 134.
This one is for utf8bm4.

Skywalker-11 replied:
With this approach we are able to first convert everything that already exists to utf8 (they should just fit in the field because the byte size of characters don't change) and then we adjust the field sizes to be able to use them as key fields. At this lines here we enable the use of utf8mb4 for future entries.

endelwar commented on upgrade.php:
```diff
@@ -174,34 +228,15 @@ function getTableIndexes($table)
-
-    $sql = 'DROP TABLE IF EXISTS `geoip_country`';
-    executeQuery($sql);
+    if (false === check_table_exists($geoip)) {
```
use geoip database name as string directly, it won't change name

stefaweb replied:
ok

stefaweb commented:
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[upgrade.php, line 157 at r7](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r7-157:-KbknE0YgeHoE7ltXn-L:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/f704f30e19939d1dfb8d0a724dbcaee24e1282d6/upgrade.php#L157)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

ok
</blockquote></details>

Done.

---

*[upgrade.php, line 180 at r7](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r7-180:-KbknFsAfA-2_Glg6SfI:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/f704f30e19939d1dfb8d0a724dbcaee24e1282d6/upgrade.php#L180)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

ok
</blockquote></details>

Done.

---

*[upgrade.php, line 184 at r7](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r7-184:-KbknGnI7Vll_7N1k90U:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/f704f30e19939d1dfb8d0a724dbcaee24e1282d6/upgrade.php#L184)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

ok
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[upgrade.php, line 234 at r7](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r7-234:-KbknLGEC5b4DSWV2ED-:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/f704f30e19939d1dfb8d0a724dbcaee24e1282d6/upgrade.php#L234)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

ok
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Question: Do you prefer the word "field" or the word "column"?

Fix schema for fullname **field** in `users` table

stefaweb commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[upgrade.php, line 83 at r9](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbpJO7-5dHY7reQWW1J:-KbpJO7-5dHY7reQWW1K:b-p3au6n) ([raw file](https://github.com/mailwatch/1.2.0/blob/7f3cba7332a1090ad6a7b4383cc546790939237d/upgrade.php#L83)):*
> ```PHP
>  * @return bool|mysqli_result
>  */
> function check_column_exists($table, $column)
> ```

Don't want to return bool !!!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

endelwar commented:
> Question: Do you prefer the word "field" or the word "column"?

I prefer field in this case

asuweb commented:
I'm a field kind of guy as well, but I suppose from a mysql point of view column is technically correct.

stefaweb commented:
ok

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-:-KbpV7AIIfGjIcWBp3J-:b32wtd)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[upgrade.php, line 90 at r10](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbqARK3D54DXHEbfGCz:-KbqARK3D54DXHEbfGD-:b-6mczhq) ([raw file](https://github.com/mailwatch/1.2.0/blob/981a7d92aed9d447cffa15e404f43afa076856ee/upgrade.php#L90)):*
> return $link->query($sql);

TO DO:

    $result = $link->query($sql);
    return mysqli_num_rows($result);

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

Skywalker-11 commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[upgrade.php, line 83 at r9](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbpJO7-5dHY7reQWW1J:-Kbqz1UH-PcCJyDfAqUF:b-yar219) ([raw file](https://github.com/mailwatch/1.2.0/blob/7f3cba7332a1090ad6a7b4383cc546790939237d/upgrade.php#L83)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

Don't want to return bool !!!
</blockquote></details>

why not `SELECT count(*) AS count FROM ...`and test `($link->query($sql)->fetch_row() )['count'] > 0 `

---

*[upgrade.php, line 90 at r10](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbqARK3D54DXHEbfGCz:-Kbr-83OimSGnSPoG4eZ:b-2q49qu) ([raw file](https://github.com/mailwatch/1.2.0/blob/981a7d92aed9d447cffa15e404f43afa076856ee/upgrade.php#L90)):*
> mysqli_num_rows 

does not exist

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[upgrade.php, line 83 at r9](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbpJO7-5dHY7reQWW1J:-KbslanMDmeO67n9Ug7Z:bwswlww) ([raw file](https://github.com/mailwatch/1.2.0/blob/7f3cba7332a1090ad6a7b4383cc546790939237d/upgrade.php#L83)):*
<details><summary><i>Previously, Skywalker-11 (Michael Pietsch) wrote…</i></summary><blockquote>

why not `SELECT count(*) AS count FROM ...`and test `($link->query($sql)->fetch_row() )['count'] > 0 `
</blockquote></details>

Added $link->query($sql)->fetch_row()

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

endelwar commented on upgrade.php:
```diff
@@ -71,11 +71,26 @@ function executeQuery($sql)
+function check_column_exists($table, $column)
+{
+    global $link;
+    $sql = 'SELECT 1 FROM information_schema.COLUMNS
```
I like more
```sql
SHOW COLUMNS FROM `$table` LIKE '$column'
```

stefaweb replied:
doesn't work

endelwar replied:
```
mysql> use mailscanner;
Database changed
mysql> SHOW COLUMNS FROM `maillog` LIKE 'subject';
+---------+------------+------+-----+---------+-------+
| Field   | Type       | Null | Key | Default | Extra |
+---------+------------+------+-----+---------+-------+
| subject | mediumtext | YES  | MUL | NULL    |       |
+---------+------------+------+-----+---------+-------+
1 row in set (0,00 sec)
```

endelwar replied:
```
mysql> SHOW COLUMNS FROM `maillog` LIKE 'idontexist';
Empty set (0,00 sec)
```

stefaweb replied:
ok, will do

endelwar commented on upgrade.php:
```diff
@@ -71,11 +71,26 @@ function executeQuery($sql)
+            AND COLUMN_NAME = "' . $column . '"
+            AND TABLE_NAME = "' . $table . '" 
+            LIMIT 1';
+    return $link->query($sql)->fetch_row();
```
This function should return true or false.
```php
return ($link->query($sql)->num_rows > 0);
```

stefaweb replied:
Doesn't work. Return alway FALSE.

```
function check_column_exists($table, $column)
{
    global $link;
    $sql = 'SHOW COLUMNS FROM `' . $table . '` LIKE "' . $table . '"';
    return ($link->query($sql)->num_rows > 0);
}

```

endelwar commented on upgrade.php:
```diff
@@ -130,36 +145,90 @@ function getTableIndexes($table)
     $sql = 'ALTER DATABASE `' . DB_NAME .
         '` CHARACTER SET = ' . $mysql_utf8_variant[$server_utf8_variant]['charset'] .
         ' COLLATE = ' . $mysql_utf8_variant[$server_utf8_variant]['collation'];
     executeQuery($sql);
```
this will also convert utf8mb4 database back to utf8

endelwar commented on upgrade.php:
```diff
@@ -71,11 +71,26 @@ function executeQuery($sql)
 {
     global $link;
-
     return $link->query('SELECT 1 FROM `' . $table . '` LIMIT 1');
```
Also this should return true of false
```sql
SHOW TABLES LIKE '$table'
```
and 
```php
return ($link->query($sql)->num_rows > 0);
```

stefaweb replied:
done

stefaweb commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[upgrade.php, line 153 at r12](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r12-153:-KbtCM3K9m7ihozpv8oc:b-iw1p49) ([raw file](https://github.com/mailwatch/1.2.0/blob/15653da8279699c8102a8f2e0768ecb7dd9e645f/upgrade.php#L153)):*
<details><summary><i>Previously, endelwar (Manuel Dalla Lana) wrote…</i></summary><blockquote>

this will also convert utf8mb4 database back to utf8
</blockquote></details>

Added test

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Not finished, not working.

stefaweb commented:
Review status: 0 of 1 files reviewed at latest revision, 12 unresolved discussions.

---

*[upgrade.php, line 104 at r14](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbtNwH5t1yb4NqyZSkg:-KbtNwH5t1yb4NqyZSkh:b-36tqhl) ([raw file](https://github.com/mailwatch/1.2.0/blob/969bccfd58370ca39ba2c312059287c792a0ebc2/upgrade.php#L104)):*
> $result = $link->query($sql);

How to get value of default_collation_name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

endelwar commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[upgrade.php, line 90 at r10](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbqARK3D54DXHEbfGCz:-KbtW8mVKVFjndLL3Fqb:b4ppgkn) ([raw file](https://github.com/mailwatch/1.2.0/blob/981a7d92aed9d447cffa15e404f43afa076856ee/upgrade.php#L90)):*
<details><summary><i>Previously, Skywalker-11 (Michael Pietsch) wrote…</i></summary><blockquote>

> mysqli_num_rows 

does not exist

</blockquote></details>

http://php.net/manual/en/mysqli-result.num-rows.php

---

*[upgrade.php, line 90 at r12](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r12-90:-KbtWLRm3v4kvbCSmMJj:bnvz4bf) ([raw file](https://github.com/mailwatch/1.2.0/blob/15653da8279699c8102a8f2e0768ecb7dd9e645f/upgrade.php#L90)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

Doesn't work. Return alway FALSE.

```
function check_column_exists($table, $column)
{
    global $link;
    $sql = 'SHOW COLUMNS FROM `' . $table . '` LIKE "' . $table . '"';
    return ($link->query($sql)->num_rows > 0);
}

```
</blockquote></details>

you don't ask for $column in LIKE part

---

*[upgrade.php, line 104 at r14](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbtNwH5t1yb4NqyZSkg:-KbtVlhrqjtzWfHrYblA:b-q0d77n) ([raw file](https://github.com/mailwatch/1.2.0/blob/969bccfd58370ca39ba2c312059287c792a0ebc2/upgrade.php#L104)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

> $result = $link->query($sql);

How to get value of default_collation_name?

</blockquote></details>

$result is a mysqli_result object, you can fetch result rows with fetch_array (http://php.net/manual/en/class.mysqli-result.php)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

endelwar commented:
Reviewed 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[upgrade.php, line 90 at r12](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r12-90:-KbtbBf4XBBa0OVNHWC_:b-2y567y) ([raw file](https://github.com/mailwatch/1.2.0/blob/15653da8279699c8102a8f2e0768ecb7dd9e645f/upgrade.php#L90)):*
<details><summary><i>Previously, endelwar (Manuel Dalla Lana) wrote…</i></summary><blockquote>

you don't ask for $column in LIKE part
</blockquote></details>

ok, see next commit

---

*[upgrade.php, line 153 at r12](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r12-153:-KbtbLEmOuglMxBy6OTj:b-s29g5p) ([raw file](https://github.com/mailwatch/1.2.0/blob/15653da8279699c8102a8f2e0768ecb7dd9e645f/upgrade.php#L153)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

Added test
</blockquote></details>

working on

---

*[upgrade.php, line 104 at r14](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbtNwH5t1yb4NqyZSkg:-Kbtb4k7Ogu3wcsJkazP:bbywahw) ([raw file](https://github.com/mailwatch/1.2.0/blob/969bccfd58370ca39ba2c312059287c792a0ebc2/upgrade.php#L104)):*
<details><summary><i>Previously, endelwar (Manuel Dalla Lana) wrote…</i></summary><blockquote>

$result is a mysqli_result object, you can fetch result rows with fetch_array (http://php.net/manual/en/class.mysqli-result.php)
</blockquote></details>

ok, will do

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r16.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[upgrade.php, line 190 at r7](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r7-190:-KbtjY9vg4N3mDjDWHI7:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/f704f30e19939d1dfb8d0a724dbcaee24e1282d6/upgrade.php#L190)):*
<details><summary><i>Previously, Skywalker-11 (Michael Pietsch) wrote…</i></summary><blockquote>

With this approach we are able to first convert everything that already exists to utf8 (they should just fit in the field because the byte size of characters don't change) and then we adjust the field sizes to be able to use them as key fields. At this lines here we enable the use of utf8mb4 for future entries.
</blockquote></details>

Done.

---

*[upgrade.php, line 153 at r12](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r12-153:-KbtjWQB0nRFVW8AykYH:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/15653da8279699c8102a8f2e0768ecb7dd9e645f/upgrade.php#L153)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

working on
</blockquote></details>

Done.

---

*[upgrade.php, line 104 at r14](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbtNwH5t1yb4NqyZSkg:-KbtjV4o60-TjhGAXuL3:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/969bccfd58370ca39ba2c312059287c792a0ebc2/upgrade.php#L104)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

ok, will do
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Reviewed 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[upgrade.php, line 90 at r10](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbqARK3D54DXHEbfGCz:-KbtqCcOeZqGWkGpbCzk:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/981a7d92aed9d447cffa15e404f43afa076856ee/upgrade.php#L90)):*
<details><summary><i>Previously, endelwar (Manuel Dalla Lana) wrote…</i></summary><blockquote>

http://php.net/manual/en/mysqli-result.num-rows.php
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[upgrade.php, line 83 at r9](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbpJO7-5dHY7reQWW1J:-KbtqGjg3FU49Y1Ba1Z-:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/7f3cba7332a1090ad6a7b4383cc546790939237d/upgrade.php#L83)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

Added $link->query($sql)->fetch_row()
</blockquote></details>

Done.

---

*[upgrade.php, line 90 at r12](https://reviewable.io:443/reviews/mailwatch/1.2.0/441#-KbMdTb01jp6QGZqEF41-r12-90:-KbtqHjQo1SzZLgxWF7B:b-896fix) ([raw file](https://github.com/mailwatch/1.2.0/blob/15653da8279699c8102a8f2e0768ecb7dd9e645f/upgrade.php#L90)):*
<details><summary><i>Previously, stefaweb (Stéphane) wrote…</i></summary><blockquote>

ok, see next commit
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

endelwar commented:
Reviewed 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mailwatch/1.2.0/441)*
<!-- Sent from Reviewable.io -->

stefaweb commented:
Not sure but I ask.

Do we really need to have the old schema in upgrade.php ?

```
 /*
    ** Updates to the schema for 1.0.3
    *
  
```

endelwar commented:
> Do we really need to have the old schema in upgrade.php ?

Well, officially the last stable version of MailWatch is 1.1.5.1, even if nowadays everyone is using some beta or rc of 1.2;
I don't remember if db schema has changed from 1.0.3 to 1.1.5.1.

I think that it's safe to remove the commented code.

stefaweb commented:
Please check schema of   autorelease

endelwar commented on upgrade.php:
```diff
@@ -130,36 +156,114 @@ function getTableIndexes($table)
+
+    // Table mcp_rules
+    echo pad(' - Fix schema for rule_desc field in `mcp_rules` table');
+    $sql = "ALTER TABLE `mcp_rules` CHANGE `rule_desc` `rule_desc` VARCHAR( 100 ) NOT NULL DEFAULT ''";
```
I can't remember why this is reduced to 100 chars from 200

stefaweb replied:
Nothing found in my email archive.

endelwar replied:
going to revert to pair with `sa_rule`
